pkg/controller/cloud/nodecontroller.go

```diff
@@ -63,96 +74,322 @@ func NewCloudNodeController(
 				}
-				// If the known node status says that Node is NotReady, then check if the node has been removed
-				// from the cloud provider. If node cannot be found in cloudprovider, then delete the node immediately
-				if currentReadyCondition != nil {
-					if currentReadyCondition.Status != v1.ConditionTrue {
-						instances, ok := cnc.cloud.Instances()
-						if !ok {
-							glog.Errorf("cloud provider does not support instances.")
-							continue
+			}
+		}
+		// If nodeIP was suggested by user, ensure that
+		// it can be found in the cloud as well (consistent with the behaviour in kubelet)
+		if nodeIP, ok := ensureNodeProvidedIPExists(node, nodeAddresses); ok {
+			if nodeIP == nil {
+				glog.Errorf("Specified Node IP not found in cloudprovider")
+				continue
+			}
+			nodeAddresses = []v1.NodeAddress{*nodeIP}
+		}
+		nodeCopy, err := api.Scheme.DeepCopy(node)
+		if err != nil {
+			glog.Errorf("failed to copy node to a new object")
+			continue
+		}
+		newNode := nodeCopy.(*v1.Node)
+		newNode.Status.Addresses = nodeAddresses
+		_, err = nodeutil.PatchNodeStatus(cnc.kubeClient, types.NodeName(node.Name), node, newNode)
+		if err != nil {
+			glog.Errorf("Error patching node with cloud ip addresses = [%v]", err)
+		}
+	}
+}
+
+// Monitor node queries the cloudprovider for non-ready nodes and deletes them
+// if they cannot be found in the cloud provider
+func (cnc *CloudNodeController) MonitorNode() {
+	instances, ok := cnc.cloud.Instances()
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
+		return
+	}
+
+	nodes, err := cnc.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
+	if err != nil {
+		glog.Errorf("Error monitoring node status: %v", err)
+		return
+	}
+
+	for i := range nodes.Items {
+		var currentReadyCondition *v1.NodeCondition
+		node := &nodes.Items[i]
+		// Try to get the current node status
+		// If node status is empty, then kubelet has not posted ready status yet. In this case, process next node
+		for rep := 0; rep < nodeStatusUpdateRetry; rep++ {
+			_, currentReadyCondition = nodeutilv1.GetNodeCondition(&node.Status, v1.NodeReady)
+			if currentReadyCondition != nil {
+				break
+			}
+			name := node.Name
+			node, err = cnc.kubeClient.CoreV1().Nodes().Get(name, metav1.GetOptions{})
+			if err != nil {
+				glog.Errorf("Failed while getting a Node to retry updating NodeStatus. Probably Node %s was deleted.", name)
+				break
+			}
+			time.Sleep(retrySleepTime)
+		}
+		if currentReadyCondition == nil {
+			glog.Errorf("Update status of Node %v from CloudNodeController exceeds retry count.", node.Name)
+			continue
+		}
+		// If the known node status says that Node is NotReady, then check if the node has been removed
+		// from the cloud provider. If node cannot be found in cloudprovider, then delete the node immediately
+		if currentReadyCondition != nil {
+			if currentReadyCondition.Status != v1.ConditionTrue {
```

Comment:
Why not combine these two conditions? eg if condition != nil && condition.status != true